File: day13/day13.cs
// Define SHOW_GAMEBOARD if running in a normal shell/console window
// and you want to see the game being played in clunky ASCII.
// Leave it undefined for a speedier result or if running from
// within the VS/VSCode debugger/console.
//#define SHOW_GAMEBOARD

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Serilog;


namespace Shunty.AdventOfCode2019
{
    public struct Location
    {
        public Int64 X;
        public Int64 Y;

        public Location(Int64 x, Int64 y)
        {
            X = x;
            Y = y;
        }
    }

    public class Tile
    {
        public Int64 X { get; set; } = 0;
        public Int64 Y { get; set; } = 0;
        public Int64 Id { get; set; } = 0;
    }

    public class Day13 : IAoCRunner
    {
        const int Empty = 0;
        const int Wall = 1;
        const int Block = 2;
        const int Paddle = 3;
        const int Ball = 4;

        // Set this to true to see a running update of the score
        // when no showing the full game window
        private bool ShowScoreUpdates = false;

        private static readonly int DayNumber = 13;

        public void Run(ILogger log)
        {
            var input = AocHelpers.GetDayText(DayNumber)
                .Split(',')
                .Select(s => Int64.Parse(s))
                .ToArray();

            var map = new Dictionary<Location, Tile>();
            var program = input.ToArray();
            RunGame(program, map);
            var grouped = map.GroupBy(m => m.Value.Id).Select(g => (g.Key, g.Count())).OrderBy(g => g.Key).ToList();
            log.Debug("Tile types: {@TileTypes}", grouped);
            var part1 = map.Count(kvp => kvp.Value.Id == Block);

            map.Clear();
            program = input.ToArray();
            program[0] = 2;
            var part2 = RunGame(program, map);

            Console.WriteLine("");
            Console.WriteLine($"Part 1: {part1}");
            Console.WriteLine($"Part 2: {part2}");
        }

        private Int64 RunGame(Int64[] program, Dictionary<Location, Tile> map)
        {
            ConcurrentQueue<Int64> inQ = new ConcurrentQueue<Int64>(), outQ = new ConcurrentQueue<Int64>();
            Int64 score = 0, paddleX = -1;

            InitGameBoard();
            InitScoreboard();
            var t = Task.Factory.StartNew(() => IntcodeCompute(program, inQ, outQ));
            int blockCount = 0;
            score = 0;
            try
            {
                while (!t.IsCompleted || outQ.Count() > 0)
                {
                    // Get 3 values from the queue
                    Int64 x = 0, y = 0, id = 0;
                    while (!outQ.TryDequeue(out x))
                    {}
                    while (!outQ.TryDequeue(out y))
                    {}
                    while (!outQ.TryDequeue(out id))
                    {}

                    if (x == -1 && y == 0)
                    {
                        score = id;
                        UpdateScore(score, blockCount);
                        if (blockCount == 0)
                            break;

                        continue;
                    }

                    if (id == Paddle)
                    {
                        if (paddleX < 0)
                        {
                            inQ.Enqueue(-1);
                        }
                        paddleX = x;
                    }
                    else if (id == Ball)
                    {
                        if (paddleX >= 0)
                        {
                            // Try and track the ball with the paddle
                            if (x < paddleX)
                                inQ.Enqueue(-1);
                            else if (x > paddleX)
                                inQ.Enqueue(1);
                            else
                                inQ.Enqueue(0);
                        }
                    }

                    var key = new Location(x, y);
                    Tile tile;
                    var delay = 10;
                    if (map.ContainsKey(key))
                    {
                        tile = map[key];
                        if (tile.Id == Block && id != Block)
                            blockCount--;
                        else if (tile.Id != Block && id == Block)
                            blockCount++;

                        tile.Id = id;
                    }
                    else
                    {
                        if (id == Block)
                            blockCount++;
                        tile = new Tile { X = x, Y = y, Id = id };
                        map[key] = tile;
                        delay = 0;
                    }

#if SHOW_GAMEBOARD
                    DrawTile(tile);
                    // Delay it a little
                    Thread.Sleep(delay);
#endif
                }

                return score;
            }
            finally
            {
                ResetScreen();
            }
        }

        private void DrawTile(Tile tile)
        {
#if !SHOW_GAMEBOARD
            return;
#else
            string s = "";
            switch ((int)tile.Id)
            {
                case Empty:
                    Console.ForegroundColor = ConsoleColor.Black;
                    s = " ";
                    break;
                case Wall:
                    Console.ForegroundColor = ConsoleColor.DarkMagenta;
                    s = tile.X == 0 ? "▌" : "▐"; //"█";
                    break;
                case Block:
                    Console.ForegroundColor = ConsoleColor.DarkBlue;
                    s = "░";
                    break;
                case Paddle:
                    Console.ForegroundColor = ConsoleColor.Blue;
                    s = "▀";
                    break;
                case Ball:
                    Console.ForegroundColor = ConsoleColor.Red;
                    s = "■";
                    break;
                default:
                    return;
            }
            Console.SetCursorPosition((int)tile.X, (int)tile.Y);
            Console.Write(s);
#endif
        }

#if SHOW_GAMEBOARD
        private int cursorTop = 0;
        private ConsoleColor orgFG = ConsoleColor.White;
        private ConsoleColor orgBG = ConsoleColor.Black;
#endif
        private void InitGameBoard()
        {
#if SHOW_GAMEBOARD
            cursorTop = Console.CursorTop;
            orgFG = Console.ForegroundColor;
            orgBG = Console.BackgroundColor;
            Console.Clear();
#endif
        }

        private void ResetScreen()
        {
#if SHOW_GAMEBOARD
            Console.ForegroundColor = orgFG;
            Console.BackgroundColor = orgBG;
            Console.SetCursorPosition(0, cursorTop);
            Console.ResetColor();
#endif
        }
        private void InitScoreboard()
        {
        }
        private void UpdateScore(Int64 score, int blockCount)
        {
#if SHOW_GAMEBOARD
            var h = Console.WindowHeight;
            Console.ForegroundColor = ConsoleColor.Green;
            Console.SetCursorPosition(0, 26);
            Console.Write($"Score: {score:#,##0};    Blocks: {blockCount:#,##0}  ");
#else
            if (ShowScoreUpdates)
                Console.WriteLine($"Score: {score}; Blocks: {blockCount}");
#endif
        }

        private Int64 ResizeProgram(ref Int64[] program, Int64 newSize)
        {
            Array.Resize(ref program, (int)newSize);
            return newSize;
        }

        private Int64 IntcodeCompute(Int64[] program, ConcurrentQueue<Int64> inQ, ConcurrentQueue<Int64> outQ)
        {
            Int64 lastoutput = 0, relativebase = 0, outaddr = 0;
            Int64 ip = 0, len = program.Length;

            while (true)
            {
                var instruction = program[ip];
                if (instruction == 99)
                    return lastoutput;

                Int64 opcode = instruction % 100;
                Int64 m1 = (instruction / 100) % 10,
                    m2 = (instruction / 1000) % 10,
                    m3 = (instruction / 10000) % 10;
                Int64 p1 = ip + 1 < len ? program[ip + 1] : 0,
                    p2 = ip + 2 < len ? program[ip + 2] : 0,
                    p3 = ip + 3 < len ? program[ip + 3] : 0;

                Int64 v1 = p1, v2 = p2;
                if (m1 == 0)
                    v1 = (p1 < len) ? program[p1] : 0;
                else if (m1 == 2)
                    v1 = (p1 + relativebase) < len ? program[p1 + relativebase] : 0;

                if (m2 == 0)
                    v2 = p2 < len ? program[p2] : 0;
                else if (m2 == 2)
                    v2 = (p2 + relativebase) < len ? program[p2 + relativebase] : 0;

                switch (opcode)
                {
                    case 1: // Add
                        outaddr = m3 == 0 ? p3 : p3 + relativebase;
                        if (outaddr >= len)
                            len = ResizeProgram(ref program, outaddr * 2);
                        program[outaddr] = v1 + v2;
                        ip += 4;
                        break;
                    case 2: // Multiply
                        outaddr = m3 == 0 ? p3 : p3 + relativebase;
                        if (outaddr >= len)
                            len = ResizeProgram(ref program, outaddr * 2);
                        program[outaddr] = v1 * v2;
                        ip += 4;
                        break;
                    case 3: // Input
                        if (inQ.TryDequeue(out var input))
                        {
                            outaddr = m1 == 0 ? p1 : p1 + relativebase;
                            if (outaddr >= len)
                                len = ResizeProgram(ref program, outaddr * 2);
                            program[outaddr] = input;
                            ip += 2;
                        }
                        break;
                    case 4: // Output
                        lastoutput = v1;
                        //Console.WriteLine($"Test result: {v1}");
                        ip += 2;
                        outQ.Enqueue(v1);
                        break;
                    case 5:  // Jump if true
                        if (v1 != 0)
                        {
                            ip = v2;
                        }
                        else
                        {
                            ip += 3;
                        }
                        break;
                    case 6:  // Jump if false
                        if (v1 == 0)
                        {
                            ip = v2;
                        }
                        else
                        {
                            ip += 3;
                        }
                        break;
                    case 7:  // Less than
                        outaddr = m3 == 0 ? p3 : p3 + relativebase;
                        if (outaddr >= len)
                            len = ResizeProgram(ref program, outaddr * 2);
                        program[outaddr] = (v1 < v2) ? 1 : 0;
                        ip += 4;
                        break;
                    case 8:  // Equals
                        outaddr = m3 == 0 ? p3 : p3 + relativebase;
                        if (outaddr >= len)
                            len = ResizeProgram(ref program, outaddr * 2);
                        program[outaddr] = (v1 == v2) ? 1 : 0;
                        ip += 4;
                        break;
                    case 9: // Update relative base
                        relativebase += v1;
                        ip += 2;
                        break;
                    case 99:
                        return lastoutput;
                    default:
                        throw new Exception($"Unknown instruction {opcode} at position {ip}");
                }
            }
       }
    }
}

File: day13/day13.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

const Block int64 = 2
const Paddle int64 = 3
const Ball int64 = 4

type location struct {
	X, Y int64
}

type tile struct {
	X, Y, TileType int64
}

func main() {
	originalInput, err := inputAsInt64Array("./day13-input.txt")
	if err != nil {
		log.Fatal(err)
	}

	input := make([]int64, len(originalInput))

	copy(input, originalInput)
	m := make(map[location]*tile)
	runGame(input, m)
	part1 := 0
	for _, t := range m {
		if t.TileType == Block {
			part1++
		}
	}
	fmt.Println("Part 1: ", part1)

	m = make(map[location]*tile)
	copy(input, originalInput)
	input[0] = 2
	part2 := runGame(input, m)
	fmt.Println("Part 2: ", part2)
}

func runGame(program []int64, m map[location]*tile) int64 {
	inQ := make(chan int64, 6)
	outQ := make(chan int64, 6)

	go runProgram(program, inQ, outQ)

	x, y, id, done, blocks := int64(0), int64(0), int64(0), false, 0
	score, paddleX := int64(0), int64(-1)

	for !done {
		// Read 3 values
		x = <-outQ
		if x == 99 {
			done = true
			break
		}
		y = <-outQ
		if y == 99 {
			done = true
			break
		}
		id = <-outQ
		if id == 99 {
			done = true
			break
		}

		// Update score ?
		if x == -1 && y == 0 {
			score = id
			if blocks == 0 {
				break
			}
			continue
		}

		if id == Paddle {
			// Initial push of the stick
			if paddleX < 0 {
				inQ <- -1 // Doesn't actually matter if -1, 0 or 1
			}
			paddleX = x
		} else if id == Ball {
			// Track the ball with the paddle
			if paddleX >= 0 {
				if x < paddleX {
					inQ <- -1
				} else if x > paddleX {
					inQ <- 1
				} else {
					inQ <- 0
				}
			}
		}

		// Update the map
		key := location{x, y}
		t, ok := m[key]
		if ok {
			if t.TileType == Block && id != Block {
				blocks++
			} else if t.TileType != Block && id == Block {
				blocks--
			}
			t.TileType = id
		} else {
			if id == Block {
				blocks++
			}
			m[key] = &tile{x, y, id}
		}
	}

	return score
}

func resizeProgram(program []int64, newSize int64) ([]int64, int64) {
	result := make([]int64, newSize)
	copy(result, program)
	return result, int64(len(result))
}

func runProgram(program []int64, inQ chan int64, outQ chan int64) (result int64, err error) {
	ip, lastoutput, len := int64(0), int64(0), int64(len(program))
	relativebase := int64(0)
	for {
		instruction := program[ip]
		if instruction == 99 {
			outQ <- int64(99)
			return lastoutput, nil
		}
		opcode := instruction % 100
		// Get the parameter modes
		m1 := (instruction / 100) % 10
		m2 := (instruction / 1000) % 10
		m3 := (instruction / 10000) % 10

		// Read the parameters - doesn't matter if we read too many for the current op
		p1, p2, p3 := int64(0), int64(0), int64(0)
		if ip+1 < len {
			p1 = program[ip+1]
		}
		if ip+2 < len {
			p2 = program[ip+2]
		}
		if ip+3 < len {
			p3 = program[ip+3]
		}

		// Set the parameter values - mode dependent
		v1, v2, v3 := p1, p2, p3
		if m1 == 0 {
			v1 = 0
			if p1 < len {
				v1 = program[p1]
			}
		} else if m1 == 2 {
			v1 = 0
			if (p1 + relativebase) < len {
				v1 = program[p1+relativebase]
			}
		}
		if m2 == 0 {
			v2 = 0
			if p2 < len {
				v2 = program[p2]
			}
		} else if m2 == 2 {
			v2 = 0
			if (p2 + relativebase) < len {
				v2 = program[p2+relativebase]
			}
		}
		if m3 == 2 {
			v3 = p3 + relativebase
		}

		switch opcode {
		case 1: // Add
			if v3 >= len {
				program, len = resizeProgram(program, v3*2)
			}
			program[v3] = v1 + v2
			ip += 4
		case 2: // Multiply
			if v3 >= len {
				program, len = resizeProgram(program, v3*2)
			}
			program[v3] = v1 * v2
			ip += 4
		case 3: // Input
			if m1 == 2 {
				p1 = p1 + relativebase
			}
			if p1 >= len {
				program, len = resizeProgram(program, p1*2)
			}
			select {
			case input := <-inQ:
				program[p1] = input
				ip += 2
			default:
				// Do nothing and just carry on
			}
		case 4: // Output
			lastoutput = v1
			//log.Println("Test result: ", v1)
			ip += 2
			outQ <- v1
		case 5: // Jump true
			if v1 != 0 {
				ip = v2
			} else {
				ip += 3
			}
		case 6: // Jump false
			if v1 == 0 {
				ip = v2
			} else {
				ip += 3
			}
		case 7: // <
			if v3 >= len {
				program, len = resizeProgram(program, v3*2)
			}

			if v1 < v2 {
				program[v3] = 1
			} else {
				program[v3] = 0
			}
			ip += 4
		case 8: // ==
			if v3 >= len {
				program, len = resizeProgram(program, v3*2)
			}
			if v1 == v2 {
				program[v3] = 1
			} else {
				program[v3] = 0
			}
			ip += 4
		case 9:
			relativebase += v1
			ip += 2
			break
		default:
			return 0, fmt.Errorf("Unexpected opcode %d", opcode)
		}
	}
}

func inputAsInt64Array(inputFile string) (nums []int64, err error) {
	txt, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return nil, err
	}

	split := strings.Split(string(txt), ",")
	nums = make([]int64, len(split))
	for i, n := range split {
		if len(n) == 0 {
			continue
		}
		num, err := strconv.Atoi(strings.Trim(n, "\r\n"))
		if err != nil {
			return nil, err
		}
		nums[i] = int64(num)
	}

	return nums, nil
}
